### PR TITLE
Parser: Allow block nesting

### PR DESCRIPTION
**Pull request by dmsnell**
In this patch we're opening up a new avenue for allowing nested blocks
in the data structure.

For each block:
 - Nested blocks appear as `innerBlocks` as a sequential list
 - The contained HTML _without_ the nested blocks appear as a string
   property `innerHtml` which replaces `rawContent` (~**update** it doesn't replace it yet~)

**Testing**

Smoke test in the browser.
Load the [grammar](https://raw.githubusercontent.com/WordPress/gutenberg/bcb4c078c227600a7a66eb80c360e8ca5f16d0fb/blocks/api/post.pegjs) into the [explorer](https://dmsnell.github.io/peg-parser-explorer/) (you can use the fetcher URL in the explorer for the grammar link) then load in some test content and play with nesting. You should be able to see how `innerBlocks` grows and when blocks do nest, that the snippet which represents them disappears from `innerHtml`.

**Input**
```html
Front

<!-- wp:core/void /-->

<!-- wp:core/nest -->
Before
<!-- wp:core/inside -->
This is working
<!-- /wp:core/inside -->
Middle
<!-- wp:core/inside -->
Another in the <!-- wp:core/mark -->jump here<!-- /wp:core/mark -->list
<!-- /wp:core/inside -->
After
<!-- /wp:core/nest -->

End
```

**Output**
```js
[
  {
    "blockName": "core/freeform",
    "rawContent": "Front\n\n"
  },
  {
    "blockName": "core/void",
    "attrs": null,
    "innerBlocks": [],
    "rawContent": ""
  },
  {
    "blockName": "core/freeform",
    "rawContent": "\n\n"
  },
  {
    "blockName": "core/nest",
    "attrs": null,
    "innerBlocks": [
      {
        "blockName": "core/inside",
        "attrs": null,
        "innerBlocks": [],
        "rawContent": "\nThis is working\n"
      },
      {
        "blockName": "core/inside",
        "attrs": null,
        "innerBlocks": [
          {
            "blockName": "core/mark",
            "attrs": null,
            "innerBlocks": [],
            "rawContent": "jump here"
          }
        ],
        "rawContent": "\nAnother in the list\n"
      }
    ],
    "rawContent": "\nBefore\n\nMiddle\n\nAfter\n"
  },
  {
    "blockName": "core/freeform",
    "rawContent": "\n\nEnd"
  }
]
```

**Notes**
 - This could end up going really slowly on account of how we are parsing a list of non-token characters and then eventually joining them.
 - I haven't tested any of this. Main purpose of the PR is to get something going to explore and see if it's good. We can update and fix all things later if we need to.

### CODE REVIEW

**Comment by BE-Webdesign** [comment]
dmsnell to the rescue! I think allowing nested blocks from the parser will be a good step forward! Awesome work so far!

Slightly off topic, but definitely tied to this, although they are more next step questions:

How do you think we will handle block normalization, or if we even want to do normalization?
How do we handle reordering of child blocks after parse ( tied to above )?

How do we parse this correctly?:
```js
{
    "blockName": "core/inside",
    "attrs": null,
    "innerBlocks": [
        {
          "blockName": "core/mark",
          "attrs": null,
          "innerBlocks": [],
          "rawContent": "jump here"
        }
    ],
    "rawContent": "\nAnother in the list\n"
}
```

"jump here" is supposed to live in between the words "the" and "list", how do we denote that location in our block data. Do we not allow other blocks within text to avoid the need for cursor data or do we need to figure out a data model that can accommodate nested blocks better, or do we create some sort of template syntax to abstract our render tree?

I don't know if this is already a solved problem, but I think the above example, in my opinion, shows that we really want to create a data spec for blocks first, and then build our API around that. Maybe with the news of changing view libraries, this could be a good time to reflect on the current data model for blocks state.

**Comment by mtias** [comment]
> "jump here" is supposed to live in between the words "the" and "list", how do we denote that location in our block data.

There are two aspects involved in determining how a block is rendered. The parsing step handled by the grammar breaks up the ingredients, but doesn't know how they are combined—that is determined by the block `edit` and `save` interfaces. Children are an _inner_ boundary of a block, so they are extracted into `innerBlocks`. From there, the block author will received as part of edit arguments `{ attributes, innerBlocks ... }`. This will be passed to a `<BlockList children={ innerBlocks } />` sort of component that would handle the UI to manage this inner list of blocks.

**Comment by dmsnell** [comment]
thanks for jumping in @BE-Webdesign (sorry I should have added you to this one).

> Awesome work so far!

much obliged 🙇 

for starters I can say sorry for pushing out a PR without much description, explanation of motivation, or context. I've been talking in person with @mtias and some others and wanted to make sure that I could record in code some of the ideas we've spoken about. will add more of the written part later as we continue to think through the implications here.

> How do you think we will handle block normalization, or if we even want to do normalization?
> How do we handle reordering of child blocks after parse ( tied to above )?
> How do we parse this correctly?:

mostly I'll just defer to the response that @mtias gave, but we had lots of discussion on this very point. this PR is the realization of the ideas we landed at.

**Comment by BE-Webdesign** [comment]
>There are two aspects involved in determining how a block is rendered. The parsing step handled by the grammar breaks up the ingredients, but doesn't know how they are combined—that is determined by the block edit and save interfaces. Children are an inner boundary of a block, so they are extracted into innerBlocks. From there, the block author will received as part of edit arguments { attributes, innerBlocks ... }. This will be passed to a <BlockList children={ innerBlocks } /> sort of component that would handle the UI to manage this inner list of blocks.

That sounds like a cool approach matias. What I am trying to figure out is how our block state will give those components enough information to render in the correct order. Maybe parsing each text node as a block would be a good solution?

**Comment by BE-Webdesign** [comment]
> (sorry I should have added you to this one)

No need to apologize. I don't expect to be added to anything. I am only a casual contributor. Handling the parsing/serialization for nested blocks will be a challenge, and hopefully I can constructively aid that discussion and path forward.

**Comment by belcherj** [comment]
@dmsnell 

- You have a merge conflict, it is much easier to resolve if you squash before rebasing
- You have a lint issue: https://github.com/belcherj/gutenberg/pull/1/commits/e1ec93f5ba2149ea2cbc2f00933e909c8ce416bb
- I think to fix the test all of the fixtures need to be updated due to the changes to the parser:  https://github.com/belcherj/gutenberg/pull/1/commits/3f31d1927e3cad158ddeec433efe27377ab6ef2d

**Comment by dmsnell** [comment]
> You have a merge conflict, it is much easier to resolve if you squash before rebasing

Thanks @belcherj - was from the work @aduth did with the `more` tag - I resolved manually from inside of GitHub

> You have a lint issue: belcherj/gutenberg@e1ec93f

Best to defer until this merges - the `innerBlocks` is there to serve as a landmark that it's available. We need to add that data into the blocks inside the app as well.

> I think to fix the test all of the fixtures need to be updated due to the changes to the parser:

Thanks! when I recreated the fixtures before they didn't all pass the tests and I didn't have the chance to dig in to find out why. Are they passing tests in your branch?

**Comment by dmsnell** [comment]
Thanks @belcherj!

Is anyone opposed to merging this PR? The _ability_ to nest blocks doesn't _force_ blocks to nest and as-is, this parser should be compliant with all existing posts.

Would anyone be interested in helping out by _testing_ this against several different existing posts?

**Comment by belcherj** [comment]
@dmsnell still a couple tests failing.  Could not quickly determine how to fix.

File: blocks/api/parser.js
Comment (aduth) [inline blocks/api/parser.js]: ~If we're truly not using `innerBlocks`, what's the intent with disabling the rule?~

Noting your comment at https://github.com/WordPress/gutenberg/pull/2743#issuecomment-333920882 , I could go either way on whether it's sensible to leave as a placeholder. It should be fine.

File: blocks/api/parser.js
Comment (aduth) [inline blocks/api/parser.js]: Should we replace all references to `rawContent` as `innerHTML` ? Here or in a separate pull changeset?

File: blocks/api/post.pegjs
Comment (aduth) [inline blocks/api/post.pegjs]: I recall we had issues previously because we don't run Babel on the `.pegjs` JavaScript, so I have a feeling this spread syntax won't work in IE11 and we'd either need to port to ES5 equivalent or investigate transpiling the result of the file.

File: blocks/api/post.pegjs
Comment (aduth) [inline blocks/api/post.pegjs]: The square bracket array syntax was added in PHP 5.4. We need to use the more-verbose `array()` for support back to PHP 5.2.4

http://php.net/manual/en/language.types.array.php

File: blocks/api/post.pegjs
Comment (aduth) [inline blocks/api/post.pegjs]: We might want to consider the capitalization here. I created #2511 a while back to open discussion on a convention. I've started leaning more toward capitalizing abbreviations (`innerHTML`) but really just care to enforce some consistency.

File: blocks/api/post.pegjs
Comment (aduth) [inline blocks/api/post.pegjs]: Anonymous functions are only available in PHP 5.3+. To support PHP 5.2.4+, we'd need this to be a named function, passing the string callable reference to `array_filter`.

http://php.net/manual/en/functions.anonymous.php

File: blocks/api/post.pegjs
Comment (aduth) [inline blocks/api/post.pegjs]: Or here we could probably do `array_filter( $children, 'is_string' );`

Comment (dmsnell) [reply blocks/api/post.pegjs]: good catch. I actually intended on maybe creating an `array_partition` function and maybe will still do that. I always get confused which core functions are _functions_ and which are _operators_ (like `empty()`) and cannot be used as `Callable`s

File: blocks/api/parser.js
Comment (dmsnell) [reply blocks/api/parser.js]: or now I thought a second PR would be good so that this can be introduced in as minimal of a way as possible

File: blocks/api/post.pegjs
Comment (dmsnell) [reply blocks/api/post.pegjs]: personally I don't care. I'll use `HTML`

Comment (dmsnell) [reply blocks/api/post.pegjs]: changed in 4e41437d

File: blocks/api/post.pegjs
Comment (dmsnell) [reply blocks/api/post.pegjs]: fixed in 4e41437d

File: blocks/api/post.pegjs
Comment (dmsnell) [reply blocks/api/post.pegjs]: fixed in 4e41437d

File: blocks/api/post.pegjs
Comment (dmsnell) [reply blocks/api/post.pegjs]: fixed in 4e41437d

File: blocks/api/post.pegjs
Comment (dmsnell) [reply blocks/api/post.pegjs]: changed in 4e41437d

**Comment by dmsnell** [comment]
All, here are some big updates since the last time I addressed this: more complete PHP support, some big comments, and a few functions extracting out specific logic for return value shapes.

I've been inspired by some alternative parser generators in how they separate actions on a rule from the grammar itself, allowing the host language to define them separately.

For example, [Canopy](http://canopy.jcoglan.com) lets you call these actions on a rule by adding something like `%combine_blocks` and then in the language-specific file you supply an actions map like this:

```js
const actions = {
	combine_blocks: function( pre, tokens, post ) {
		…
	}
}
```

Anyway, something to think about as we start introducing more code. I'd love to have a grammar completely free from any code.

File: blocks/api/post.pegjs
Comment (aduth) [inline blocks/api/post.pegjs]: Elsewhere we have made assumptions that freeform is the absence of a block name. Specifically, [this test case](https://github.com/WordPress/gutenberg/blob/6e984b0a13392bd85159eccc989bd56d4cdd0779/blocks/api/test/parser.js#L381-L416) now fails for a few reasons:

- In block registration we allow unknown type handler to be assigned to any block name, not assuming `core/freeform`
- The content of the freeform block now includes the block delimiters `<!-- wp:freeform -->` which was not the case previously and arguably not desirable (see also #2513)

File: blocks/api/post.pegjs
Comment (aduth) [inline blocks/api/post.pegjs]: ❤️  A good place for micro-optimizations.

File: blocks/api/post.pegjs
Comment (aduth) [inline blocks/api/post.pegjs]: Unnecessary semi-colon. Guessing it wouldn't be very easy for us to apply linting to this file, would it?

**Review by aduth — COMMENTED** [review]
Looks like fixtures need to be regenerated per `innerHtml` -> `innerHTML` change:

```
npm run fixtures:regenerate
```

Resolves some, but not all PHPUnit errors.

It looks like we no longer guarantee that `attrs` will exist on the returned parsed block object, so there are PHP errors occurring here where we assume to be able to access the property:

https://github.com/WordPress/gutenberg/blob/6e984b0a13392bd85159eccc989bd56d4cdd0779/lib/blocks.php#L77

```
1) Dynamic_Blocks_Render_Test::test_dynamic_block_rendering
Undefined index: attrs
```

File: blocks/api/post.pegjs
Comment (dmsnell) [reply blocks/api/post.pegjs]: is it? I thought I would be following style guidelines by including it.

Comment (aduth) [reply blocks/api/post.pegjs]: Yes, unnecessary: http://bit.ly/2y5BO9c

**Comment by dmsnell** [comment]
Thanks @aduth - I have updated with your feedback!

File: blocks/api/parser.js
Comment (aduth) [inline blocks/api/parser.js]: We probably want to normalize this toward `innerHTML` and drop `rawContent`, for consistency, and also because objects of `innerBlocks` will keep their `innerHTML` property. Should be fine to do separately.

**Review by aduth — APPROVED** [review]

**Review by aduth — CHANGES_REQUESTED** [review]
Oops, looks like there's still an issue: Support for passing strings as callables in PHP was only added in PHP 5.4, which does not meet our support requirements of 5.2.4:

http://php.net/manual/en/language.types.callable.php
https://wordpress.org/about/requirements/

**Review by aduth — APPROVED** [review]

**Comment by aduth** [comment]
It was easy enough to drop the callable type hint and have it work just as well. See c344da6.

I'd personally like to see us push forward with block nesting. It would help in other cases such as #3047 and #2854 where I think we've had to make some compromises due to lack of proper nesting support.

**Comment by dmsnell** [comment]
@aduth what’s the point of removing the type hint?

**Comment by aduth** [comment]
@dmsnell See https://github.com/WordPress/gutenberg/pull/2743#pullrequestreview-72337655

**Comment by dmsnell** [comment]
oh roger @aduth - I thought that the act itself of passing a string callable was the issue, not adding the type hint

**Comment by dmsnell** [comment]
Thanks for all your help on this one @aduth!

**Comment by nylen** [comment]
There are no test cases for this major and foundational change to the parser, even though there is a framework in place for creating them.

What is the plan for adding these test cases?

**Pull request merged**
